File: atCoderEnv/problems/tessoku-book/a77/a77.py
from sys import stdin, setrecursionlimit

setrecursionlimit(10**6)


def main():
    N, L = map(int, stdin.readline().split())
    K = int(stdin.readline())
    *A, = map(int, stdin.readline().split())

    def check(x):
        count, last_kireme = 0, 0
        for i in range(N):
            if A[i] - last_kireme >= x and L - A[i] >= x:
                count += 1
                last_kireme = A[i]
        if count >= K:
            return True
        else:
            return False

    left = 0
    right = 1000000000
    while left < right:
        mid = (left+right+1)//2
        ans = check(mid)
        if ans == False:
            right = mid - 1
        else:
            left = mid

    print(left)


if __name__ == '__main__':
    main()

File: atCoderEnv/problems/tessoku-book/a77/a77.cpp
#include <bits/stdc++.h>
using namespace std;
#include <atcoder/all>
using namespace atcoder;
/* alias */
using ull = unsigned long long;
using ll = long long;
using vi = vector<int>;
using vvi = vector<vi>;  // 二次元vector
using vs = vector<string>;
using pii = pair<int, int>;
/* macro */
#define rep(i, a, n) for (int i = (a); i < (n); ++i)

int N, L, K, A[100009];

bool check(int x) {
  int cnt = 0, last_kireme = 0;
  rep(i, 1, N + 1) {
    if (A[i] - last_kireme >= x && L - A[i] >= x) {
      cnt += 1;
      last_kireme = A[i];
    }
  }
  return cnt >= K;
}

int main() {
  cin >> N >> L >> K;
  rep(i, 1, N + 1) cin >> A[i];
  int left = 1, right = 1000000000;
  while (left < right) {
    int mid = (left + right + 1) / 2;
    bool ans = check(mid);
    if (ans) {
      left = mid;
    } else {
      right = mid - 1;
    }
  }
  cout << left << endl;
  return 0;
}